src/rust/ide/lib/view/src/graph_editor/component/project_name.rs

```diff
@@ -0,0 +1,314 @@
+        let view_logger           = Logger::sub(&logger,"view_logger");
+        let view                  = component::ShapeView::<shape::Shape>::new(&view_logger,scene);
+        let project_name          = Rc::new(RefCell::new(UNKNOWN_PROJECT_NAME.to_string()));
+        let renamed_output        = frp.outputs.renamed.clone();
+        let animations            = Animations::new(&frp.network);
+        Self{logger,view,display_object,text_field,project_name,renamed_output
+            ,animations}.initialize()
+    }
+
+    fn setup_center_alignment(&self) {
+        let mut width = 0.0;
+        self.text_field.with_mut_content(|content| {
+            let mut line = content.line(0);
+            if line.len() > 0 {
+                width = line.get_char_x_position(line.len() - 1);
+            }
+        });
+        let offset = Vector3::new(-width/2.0,0.0,0.0);
+        let mut height = default();
+            self.text_field.with_content(|content| height = content.line_height);
+        self.view.shape.sprite.size.set(Vector2::new(width,height));
+        self.view.set_position(Vector3::new(0.0,-height/2.0,0.0));
+        self.animations.positioning.set_target_value(offset);
+    }
+
+    fn initialize(self) -> Self {
+        self.add_child(&self.text_field.display_object());
+        self.add_child(&self.view);
+        let project_name = self.clone_ref();
+        self.text_field.set_text_edit_callback(move |change| {
+            // If the text edit callback is called, the TextEdit must be still alive.
+            let field_content = project_name.text_field.get_content();
+            let new_name      = field_content.replace("\n","");
+            if change.inserted == "\n" {
+                project_name.rename(&new_name);
+            }
+            // Keep only one line.
+            project_name.text_field.set_content(&new_name);
+            project_name.setup_center_alignment();
+        });
+
+        self.setup_text_field_content();
+        self
+    }
+
+    fn reset_name(&self) {
+        info!(self.logger, "Resetting project name.");
+        self.setup_text_field_content();
+    }
+
+    fn setup_text_field_content(&self) {
+        self.text_field.set_content(&self.project_name.borrow());
+        self.setup_center_alignment();
+    }
+
+    fn highlight_text(&self) {
+        self.animations.highlight.set_target_value(1.0);
+    }
+
+    fn darken_text(&self) {
+        if !self.text_field.is_focused() {
+            self.animations.highlight.set_target_value(0.0);
+        }
+    }
+
+    fn set_highlight(&self, value:f32) {
+        let base_color = linear_interpolation(DARK_GRAY_TEXT_COLOR,HIGHLIGHTED_TEXT_COLOR,value);
+        self.text_field.set_base_color(base_color);
+    }
+
+    fn set_position(&self, value:Vector3<f32>) {
+        self.text_field.set_position(value);
+    }
+
+    fn rename(&self, name:impl Str) {
+        let name = name.into();
+        *self.project_name.borrow_mut() = name.clone();
```

Comment:
why `name.clone()`?  Whydo you want to allocate if for example the argument was a string passed as ownership? Use `name.into()` - will be moreefficient

Reply:
Because I need to move it in the 4th line:
```rust
        let name = name.into();
        *self.project_name.borrow_mut() = name.clone();
        self.update_text_field_content();
        self.name_output.emit(name);
```

Reply:
That's not true at all. `emit` accepts reference

Reply:
As a side note: this is a drawback of the philosophy of getting "AsRef/Into" everywhere - it is not clear when function can take reference, or need to take the whole value. If the function would just take a reference instead of impl ToRef, the compiler would point here, that functions needs reference only and it would be clear that clone is not needed here.

Reply:
@farmaazon very interesting observation indeed.

Reply:
The original issue was fixed.